ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantSessionManager.scala

```diff
@@ -3,26 +3,62 @@
 
 package com.daml.ledger.api.testtool.infrastructure.participant
 
+import com.daml.ledger.api.testtool.infrastructure.participant.ParticipantSessionManager._
+import io.grpc.ManagedChannel
 import io.grpc.netty.{NegotiationType, NettyChannelBuilder}
+import io.netty.channel.EventLoopGroup
 import io.netty.channel.nio.NioEventLoopGroup
 import io.netty.channel.socket.nio.NioSocketChannel
 import io.netty.util.concurrent.DefaultThreadFactory
 import org.slf4j.LoggerFactory
 
-import scala.collection.concurrent.TrieMap
+import scala.collection.immutable
+import scala.concurrent.duration.SECONDS
 import scala.concurrent.{ExecutionContext, Future}
 
-private[infrastructure] final class ParticipantSessionManager {
+private[infrastructure] final class ParticipantSessionManager(
+    sessions: immutable.Map[ParticipantSessionConfiguration, SessionParts],
+) {
+  lazy val all: immutable.Seq[ParticipantSession] = sessions.values.toVector.map(_._1)
 
-  private[this] val logger = LoggerFactory.getLogger(classOf[ParticipantSession])
+  def get(configuration: ParticipantSessionConfiguration): ParticipantSession =
+    sessions(configuration)._1
 
-  private[this] val channels = TrieMap.empty[ParticipantSessionConfiguration, ParticipantSession]
+  def closeAll(): Unit =
+    for ((config, (_, channel, eventLoopGroup)) <- sessions) {
+      logger.info(s"Disconnecting from participant at ${config.address}...")
+      channel.shutdownNow()
+      if (!channel.awaitTermination(10L, SECONDS)) {
+        sys.error("Channel shutdown stuck. Unable to recover. Terminating.")
+      }
+      logger.info(s"Connection to participant at ${config.address} shut down.")
+      if (!eventLoopGroup
+          .shutdownGracefully(0, 0, SECONDS)
+          .await(10L, SECONDS)) {
+        sys.error("Unable to shutdown event loop. Unable to recover. Terminating.")
+      }
+      logger.info(s"Connection to participant at ${config.address} closed.")
+    }
+}
+
+object ParticipantSessionManager {
+  private type SessionParts = (ParticipantSession, ManagedChannel, EventLoopGroup)
+
+  private val logger = LoggerFactory.getLogger(classOf[ParticipantSession])
+
+  def apply(configs: immutable.Seq[ParticipantSessionConfiguration])(
+      implicit executionContext: ExecutionContext
+  ): Future[ParticipantSessionManager] =
+    for {
+      participantSessions <- Future
+        .traverse(configs)(config => connect(config).map(config -> _))
+        .map(_.toMap)
+    } yield new ParticipantSessionManager(participantSessions)
```

Comment:
```suggestion
    } yield new ParticipantSessionManager(participantSessions.toMap)
```
This should be equivalent, right?

Reply:
It is, well spotted. Thanks!